### DISCUSSION TITLE
How to efficiently swap bind group resources?

### DISCUSSION THREAD

**Discussion by erlendp** [gfx-rs/wgpu#1495] [Q&A] [answered]
Does anyone know what the most efficient way of swapping bind group resources would be? I'm looking to swap out a texture resource in a bind group. Aside from updating the existing resource in GPU memory with new data, I assume I'll need to create a new bind group any time I want to change the bound texture? If so, is there any plan to provide a method for updating existing bindings in the future (aka vkUpdateDescriptorSets)?

**Comment by erlendp** [comment, marked as answer]
Looks like I may have found an answer - https://github.com/gpuweb/gpuweb/issues/915.

Long story short, it sounds like the overhead to creating bind groups on the fly should be minimal, as the implementation should be able to use some kind of caching under the hood to achieve this via `vkUpdateDescriptorSets` (in the case of Vulkan), although I'm unclear as to whether this implementation actually does.

Also sounds like there's a proposal to introduce a `BindGroupArena` to manage these kinds of dynamic bind groups - https://github.com/gfx-rs/wgpu/issues/888

**Reply by cwfitzgerald** [reply]
We don't do any kind of caching of bind groups. The only caching we do is of bind group layouts. Generally speaking wgpu tries to be pretty dumb when it comes to deduplication because we want to keep overhead as low as possible.

Currently the only way to update a bind group resource is to create an entirely new bind group. BG creation isn't _that_ expensive especially for small BG's, so it shouldn't be too much of a problem.
